htmresearch/frameworks/layers/l2_l4_network_creation.py

```diff
@@ -303,12 +303,21 @@ def createMultipleL4L2Columns(network, networkConfig):
   for i in range(networkConfig["numCorticalColumns"]):
     suffixSrc = "_" + str(i)
     for j in range(networkConfig["numCorticalColumns"]):
-      if i != j:
-        suffixDest = "_" + str(j)
+      suffixDest = "_" + str(j)
+      if i < j:
+        # use delayed link for lateral connections because c_i is computed
+        # before c_j
         network.link(
-            "L2Column" + suffixSrc, "L2Column" + suffixDest,
-            "UniformLink", "",
-            srcOutput="feedForwardOutput", destInput="lateralInput")
+          "L2Column" + suffixSrc, "L2Column" + suffixDest,
+          "UniformLink", "",
+          srcOutput="feedForwardOutput", destInput="lateralInput",
+          propagationDelay=1)
+      elif i > j:
+        network.link(
+          "L2Column" + suffixSrc, "L2Column" + suffixDest,
+          "UniformLink", "",
+          srcOutput="feedForwardOutput", destInput="lateralInput",
+          propagationDelay=0)
```

Comment:
Yes, it is necessary. A delay on the link from column i to column j is only appropriate If column i is computed before column j. 

One way to avoid this check is to change the logic in Network.cpp
https://github.com/numenta/nupic.core/blob/master/src/nupic/engine/Network.cpp#L442

We could first prepareInputs for all regions that has the same phase, and then do the compute step. I feel changing that would affect many other things though.